src/Xamarin.Android.Build.Tasks/Utilities/LlvmIrGenerator/FunctionAttributes.cs

```diff
@@ -0,0 +1,793 @@
+using System;
+using System.Text;
+
+namespace Xamarin.Android.Tasks.LLVMIR
+{
+	// Not all attributes are currently used throughout the code, but we define them call for potential future use.
+	// Documentation can be found here: https://llvm.org/docs/LangRef.html#function-attributes
+	abstract class LLVMFunctionAttribute
+	{
+		public string Name { get; }
+		public bool Quoted { get; }
+		public bool SupportsParams { get; }
+		public bool ParamsAreOptional { get; }
+		public bool HasValueAsignment { get; }
+
+		protected LLVMFunctionAttribute (string name, bool quoted, bool supportsParams, bool optionalParams, bool hasValueAssignment)
+		{
+			Name = EnsureNonEmptyParameter (nameof (name), name);
+
+			if (supportsParams && hasValueAssignment) {
+				throw new InvalidOperationException ($"Function attribute '{name}' cannot have both parameters and an assigned value");
+			}
+
+			ParamsAreOptional = optionalParams;
+			SupportsParams = supportsParams;
+			HasValueAsignment = hasValueAssignment;
+			Quoted = quoted;
+		}
+
+		public string Render ()
+		{
+			var sb = new StringBuilder ();
+
+			if (Quoted) {
+				sb.Append ('"');
+			}
+
+			sb.Append (Name);
+
+			if (Quoted) {
+				sb.Append ('"');
+			}
+
+			if (SupportsParams) {
+				if (!ParamsAreOptional || HasOptionalParams ()) {
+					sb.Append ('(');
+					RenderParams (sb);
+					sb.Append (')');
+				}
+			} else if (HasValueAsignment) {
+				sb.Append ('=');
+				if (Quoted) {
+					sb.Append ('"');
+				}
+
+				RenderAssignedValue (sb);
```

Comment:
Semantically, it feels like `RenderAssignedValue()` should ensure that the assigned value does *not* contain `"`, lest it prematurely end the quoting.  However, none of the overrides do anything to escape or ensure that the value printed doesn't contain `"`; see e.g. `TuneCpuFunctionAttribute`.

Currently, this might not be a problem, but eventually the object graph will be "too big" to easily audit all locations to *know* that `"` won't ever be present in a value.

Should a method be added to deal with escaping?

```csharp
protected void AppendValue (StringBuilder sb, string value)
{
    sb.Append (value.Replace ("\"", "\\\"");
}
```

Reply:
I didn't see any values with quotes in the LLVM IR docs, but I will make that change